File: leetcode/python/543_diameter_of_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def __init__(self) -> None:
        self.cache: dict[TreeNode, int] = {}

    def diameterOfBinaryTree(self, root: TreeNode | None) -> int:
        if not root:
            return 0
        l = self.getDepth(root.left)
        r = self.getDepth(root.right)
        return max(l+r, self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right))
    
    def getDepth(self, root: TreeNode | None) -> int:
        if not root:
            return 0
        if root in self.cache:
            return self.cache[root]
        l = self.getDepth(root.left)
        r = self.getDepth(root.right)
        self.cache[root] = max(l, r) + 1
        return self.cache[root]

File: leetcode/rust/src/543_diameter_of_binary_tree.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut max = 0;
        Self::depth(&root, &mut max);
        max
    }

    fn depth(root: &Option<Rc<RefCell<TreeNode>>>, max: &mut i32) -> i32 {
        if let Some(node) = root {
            let node = node.borrow();
            let left = Self::depth(&node.left, max);
            let right = Self::depth(&node.right, max);
            *max = std::cmp::max(*max, left + right);
            std::cmp::max(left, right) + 1
        } else {
            0
        }
    }
}